### ISSUE TITLE
Assert that entries to softmax are appropriate

### ISSUE THREAD

**Issue by cemoody** [pfnet/chainer#302]
No single entry in the input array of the softmax should exceed that shape of the input array.

So for example,  if you call `F.softmax(y)` then `assert y.max() < y.shape[1]` should be True. This does introduce some overhead, so perhaps is worth debating -- but I just spent a few hours debugging such a bug so I'm in favor :)

**Comment by beam2d** [comment]
That would be very helpful for debugging. In order to keep the maximum speed, we might have to think about introducing _debug mode_ to diagnose such semantic errors with additional overheads.
